### PR TITLE
feat(upgrade, jiva): add runtasks and job to upgrade jiva volumes from 0.8.2 to 0.9.0

### PR DESCRIPTION
**Pull request by mittachaitu**
Signed-off-by: mittachaitu <sai.chaithanya@mayadata.io>

This PR adds support m-upgrade for jiva volumes from 0.8.2 to 0.9.0.

**Steps to upgrade:**
```
step1: git clone https://github.com/openebs/openebs.git
step2: cd openebs/k8s/upgrades/0.8.2-0.9.0/
step3: Update the `volume-upgrade-job.yaml` with corresponding jiva pv name and namespace.
step1: kubectl apply -f rbac.yaml
step2: kubectl apply -f upgrade-job.yaml
step3: kubectl apply -f jiva_upgrade_runtask.yaml
step4: kubectl apply -f volume-upgrade-job.yaml
```
**sample output:**
```
kubectl describe upgraderesult
Name:         volume-upgrade-5zcsq
Namespace:    default
Labels:       upgradeitem.openebs.io/kind=jiva-volume
              upgradeitem.openebs.io/name=pvc-58654bd9-7861-11e9-b212-54e1ad4a9dd4
              upgradeitem.openebs.io/namespace=default
              upgradejob.openebs.io/name=volume-upgrade
Annotations:  <none>
API Version:  openebs.io/v1alpha1
Config:
  API Version:  
  Data:         <nil>
  Generation:   
  Kind:         jiva-volume
  Name:         pvc-58654bd9-7861-11e9-b212-54e1ad4a9dd4
  Namespace:    default
Kind:           UpgradeResult
Metadata:
  Creation Timestamp:  2019-05-17T05:06:02Z
  Generate Name:       volume-upgrade-
  Generation:          1
  Owner References:
    API Version:           batch/v1
    Block Owner Deletion:  true
    Controller:            true
    Kind:                  Job
    Name:                  volume-upgrade
    UID:                   748ba169-7861-11e9-b212-54e1ad4a9dd4
  Resource Version:        3100
  Self Link:               /apis/openebs.io/v1alpha1/namespaces/default/upgraderesults/volume-upgrade-5zcsq
  UID:                     77d64a3d-7861-11e9-b212-54e1ad4a9dd4
Status:
  Actual Count:   0
  Desired Count:  0
  Failed Count:   0
  Resource:
    API Version:  
    Generation:   
    Kind:         <no value>
    Name:         <no value>
    Namespace:    <no value>
    Post State:
      Last Transition Time:  <nil>
      Message:               
      Status:                
    Pre State:
      Last Transition Time:  <nil>
      Message:               
      Status:                
    Sub Resources:           <nil>
Tasks:
  End Time:              2019-05-17T05:06:04Z
  Last Error:            
  Last Transition Time:  <nil>
  Message:               patched UpgradeResult {volume-upgrade-5zcsq} with name and namespace of the resource to be upgraded
  Name:                  upgrade-jiva-volume-0.8.2-0.9.0-patch-upgrade-results
  Retries:               0
  Start Time:            2019-05-17T05:06:04Z
  Status:                Success
  End Time:              2019-05-17T05:06:04Z
  Last Error:            
  Last Transition Time:  <nil>
  Message:               details of volume {pvc-58654bd9-7861-11e9-b212-54e1ad4a9dd4}: volCASType: {jiva}, storageClassName: {jiva-1r-raa}, pvcName: {jiva-vol1-1r-raa-claim}, pvcNamespace: {default}
  Name:                  upgrade-jiva-volume-0.8.2-0.9.0-get-volume-details
  Retries:               0
  Start Time:            2019-05-17T05:06:04Z
  Status:                Success
  End Time:              2019-05-17T05:06:04Z
  Last Error:            
  Last Transition Time:  <nil>
  Message:               resource version for StorageClass {jiva-1r-raa}: {2811}
  Name:                  upgrade-jiva-volume-0.8.2-0.9.0-get-sc-res-version
  Retries:               0
  Start Time:            2019-05-17T05:06:04Z
  Status:                Success
  End Time:              2019-05-17T05:06:05Z
  Last Error:            
  Last Transition Time:  <nil>
  Message:               target deployment: {pvc-58654bd9-7861-11e9-b212-54e1ad4a9dd4-ctrl} is in expected version
  Name:                  upgrade-jiva-volume-0.8.2-0.9.0-get-list-ctrl-deployment
  Retries:               0
  Start Time:            2019-05-17T05:06:05Z
  Status:                Success
  End Time:              2019-05-17T05:06:05Z
  Last Error:            
  Last Transition Time:  <nil>
  Message:               replica deployment: {pvc-58654bd9-7861-11e9-b212-54e1ad4a9dd4-rep} is in expected version
  Name:                  upgrade-jiva-volume-0.8.2-0.9.0-get-list-rep-deployment
  Retries:               0
  Start Time:            2019-05-17T05:06:05Z
  Status:                Success
  End Time:              2019-05-17T05:06:05Z
  Last Error:            
  Last Transition Time:  <nil>
  Message:               target service: {pvc-58654bd9-7861-11e9-b212-54e1ad4a9dd4-ctrl-svc} is in expected version
  Name:                  upgrade-jiva-volume-0.8.2-0.9.0-get-list-ctrl-svc
  Retries:               0
  Start Time:            2019-05-17T05:06:05Z
  Status:                Success
  End Time:              2019-05-17T05:06:05Z
  Last Error:            
  Last Transition Time:  <nil>
  Message:               replicaset to be deleted after patching target deployment {pvc-58654bd9-7861-11e9-b212-54e1ad4a9dd4-ctrl} is : {pvc-58654bd9-7861-11e9-b212-54e1ad4a9dd4-ctrl-7b8ddcd686}
  Name:                  upgrade-jiva-volume-0.8.2-0.9.0-get-list-ctrl-old-rs
  Retries:               0
  Start Time:            2019-05-17T05:06:05Z
  Status:                Success
  End Time:              2019-05-17T05:06:05Z
  Last Error:            
  Last Transition Time:  <nil>
  Message:               replicasets to be deleted after patching replica deployment: {pvc-58654bd9-7861-11e9-b212-54e1ad4a9dd4-rep} is : {pvc-58654bd9-7861-11e9-b212-54e1ad4a9dd4-rep-546c95f6b9}
  Name:                  upgrade-jiva-volume-0.8.2-0.9.0-get-list-rep-old-rs
  Retries:               0
  Start Time:            2019-05-17T05:06:05Z
  Status:                Success
  End Time:              2019-05-17T05:06:05Z
  Last Error:            
  Last Transition Time:  <nil>
  Message:               replica nodeNames for volume: {pvc-58654bd9-7861-11e9-b212-54e1ad4a9dd4} are {127.0.0.1}
  Name:                  upgrade-jiva-volume-0.8.2-0.9.0-get-list-rep-nodenames
  Retries:               0
  Start Time:            2019-05-17T05:06:05Z
  Status:                Success
  End Time:              2019-05-17T05:06:05Z
  Last Error:            
  Last Transition Time:  <nil>
  Message:               controller deployment: {pvc-58654bd9-7861-11e9-b212-54e1ad4a9dd4-ctrl} patched with latest images version: {v0.9.x-ci}
  Name:                  upgrade-jiva-volume-0.8.2-0.9.0-patch-ctrl-deployment-latest-version
  Retries:               0
  Start Time:            2019-05-17T05:06:05Z
  Status:                Success
  End Time:              2019-05-17T05:06:05Z
  Last Error:            
  Last Transition Time:  <nil>
  Message:               target deployment: {pvc-58654bd9-7861-11e9-b212-54e1ad4a9dd4-ctrl} rollout status: success
  Name:                  upgrade-jiva-volume-0.8.2-0.9.0-post-check-ctrl-deployment-status-latest-version
  Retries:               0
  Start Time:            2019-05-17T05:06:05Z
  Status:                Success
  End Time:              2019-05-17T05:06:05Z
  Last Error:            
  Last Transition Time:  <nil>
  Message:               patched controller deployment: {pvc-58654bd9-7861-11e9-b212-54e1ad4a9dd4-ctrl} with latest images version: {v0.9.x-ci}
  Name:                  upgrade-jiva-volume-0.8.2-0.9.0-post-check-ctrl-deployment-image
  Retries:               0
  Start Time:            2019-05-17T05:06:05Z
  Status:                Success
  End Time:              2019-05-17T05:06:06Z
  Last Error:            
  Last Transition Time:  <nil>
  Message:               replica deployment: {pvc-58654bd9-7861-11e9-b212-54e1ad4a9dd4-rep} patched with latest images version: {v0.9.x-ci}
  Name:                  upgrade-jiva-volume-0.8.2-0.9.0-patch-rep-deployment-latest-version
  Retries:               0
  Start Time:            2019-05-17T05:06:06Z
  Status:                Success
  End Time:              2019-05-17T05:06:33Z
  Last Error:            
  Last Transition Time:  <nil>
  Message:               replica deployment: {pvc-58654bd9-7861-11e9-b212-54e1ad4a9dd4-rep} rollout status: success
  Name:                  upgrade-jiva-volume-0.8.2-0.9.0-post-check-rep-deployment-status-latest-version
  Retries:               0
  Start Time:            2019-05-17T05:06:33Z
  Status:                Success
  End Time:              2019-05-17T05:06:33Z
  Last Error:            
  Last Transition Time:  <nil>
  Message:               patched version label on controller service: {pvc-58654bd9-7861-11e9-b212-54e1ad4a9dd4-ctrl-svc}
  Name:                  upgrade-jiva-volume-0.8.2-0.9.0-post-check-rep-deployment-image
  Retries:               0
  Start Time:            2019-05-17T05:06:33Z
  Status:                Success
  End Time:              2019-05-17T05:06:33Z
  Last Error:            
  Last Transition Time:  <nil>
  Message:               patched controller service: {pvc-58654bd9-7861-11e9-b212-54e1ad4a9dd4-ctrl-svc} with required labels and annotations
  Name:                  upgrade-jiva-volume-0.8.2-0.9.0-patch-ctrl-svc
  Retries:               0
  Start Time:            2019-05-17T05:06:33Z
  Status:                Success
  End Time:              <nil>
  Last Error:            
  Last Transition Time:  <nil>
  Message:               
  Name:                  upgrade-jiva-volume-0.8.2-0.9.0-post-check-ctrl-svc
  Retries:               0
  Start Time:            <nil>
  Status:                
  End Time:              2019-05-17T05:06:33Z
  Last Error:            
  Last Transition Time:  <nil>
  Message:               deleted older controller replicasets: {pvc-58654bd9-7861-11e9-b212-54e1ad4a9dd4-ctrl-7b8ddcd686}
  Name:                  upgrade-jiva-volume-0.8.2-0.9.0-delete-old-ctrl-rs
  Retries:               0
  Start Time:            2019-05-17T05:06:33Z
  Status:                Success
  End Time:              2019-05-17T05:06:34Z
  Last Error:            
  Last Transition Time:  <nil>
  Message:               deleted older replica replicasets: {pvc-58654bd9-7861-11e9-b212-54e1ad4a9dd4-ctrl-7b8ddcd686}
  Name:                  upgrade-jiva-volume-0.8.2-0.9.0-delete-old-rep-rs
  Retries:               0
  Start Time:            2019-05-17T05:06:34Z
  Status:                Success
Events:                  <none>

```

### CODE REVIEW

**Comment by utkarshmani1997** [comment]
@mittachaitu please move these files to `jiva` dir in the same dir.

File: k8s/upgrades/0.8.2-0.9.0/jiva_upgrade_runtask.yaml [hunk 1]
```diff
@@ -0,0 +1,668 @@
...
+  task: |-
+    type: merge
+    pspec: |-
+      status:
+        resource:
+          name: {{ .upgradeItem.name }}
+          namespace: {{ .upgradeItem.namespace }}
+          kind: {{ .upgradeItem.kind }}
+  post: |
+    {{- $message := printf "upgradeResult {%s} has been patched with basic details such as name and namespace of the resource to be upgraded." .UpgradeItem.upgradeResultName -}}
```
[the remark this answers is not in the document]
Comment (mittachaitu) [reply k8s/upgrades/0.8.2-0.9.0/jiva_upgrade_runtask.yaml]: Done

[the remark this answers is not in the document]
Comment (mittachaitu) [reply k8s/upgrades/0.8.2-0.9.0/jiva_upgrade_runtask.yaml]: Done

File: k8s/upgrades/0.8.2-0.9.0/jiva_upgrade_runtask.yaml [hunk 2]
```diff
@@ -0,0 +1,668 @@
...
+  meta: |
+    id: getvoldetails
+    apiVersion: v1
+    kind: PersistentVolume
+    action: get
+    objectName: {{ .UpgradeItem.name }}
+    runNamespace: {{ .UpgradeItem.namespace }}
+  post: |
+    {{- jsonpath .JsonResult "{.metadata.annotations.openebs\\.io/cas-type}" | trim | saveAs "getvoldetails.volCASType" .TaskResult | noop -}}
+    {{- .TaskResult.getvoldetails.volCASType | notFoundErr "volume CAS type not found" | saveIf "getvoldetails.notFoundErr" .TaskResult | noop -}}
```
[the remark this answers is not in the document]
Comment (mittachaitu) [reply k8s/upgrades/0.8.2-0.9.0/jiva_upgrade_runtask.yaml]: Done but still can be improved

File: k8s/upgrades/0.8.2-0.9.0/jiva_upgrade_runtask.yaml [hunk 3]
```diff
@@ -0,0 +1,668 @@
...
+    {{- jsonpath .JsonResult "{.spec.storageClassName}" | trim | saveAs "getvoldetails.storageClassName" .TaskResult | noop -}}
```
[the remark this answers is not in the document]
Comment (mittachaitu) [reply k8s/upgrades/0.8.2-0.9.0/jiva_upgrade_runtask.yaml]: Done

File: k8s/upgrades/0.8.2-0.9.0/jiva_upgrade_runtask.yaml [hunk 4]
```diff
@@ -0,0 +1,668 @@
...
+    {{- $status := .Config.successStatus.value -}}
+
+    {{- $URName := upgradeResultWithTaskOwnerName .UpgradeItem.upgradeResultName -}}
+    {{- $URNamespace := upgradeResultWithTaskOwnerNamespace "default" -}}
+    {{- $taskName := upgradeResultWithTaskName "upgrade-jiva-volume-0.8.2-0.9.0-get-volume-details" -}}
+    {{- $taskStatus := upgradeResultWithTaskStatus $status -}}
+    {{- $taskMessage := upgradeResultWithTaskMessage $message -}}
+    {{- $taskStartTime := upgradeResultWithTaskStartTime now -}}
+    {{- $taskEndTime := upgradeResultWithTaskEndTime now -}}
+    #    {{- $taskRetries := upgradeResultWithTaskRetries 7 -}}
```
[the remark this answers is not in the document]
Comment (mittachaitu) [reply k8s/upgrades/0.8.2-0.9.0/jiva_upgrade_runtask.yaml]: Done

File: k8s/upgrades/0.8.2-0.9.0/jiva_upgrade_runtask.yaml [hunk 5]
```diff
@@ -0,0 +1,668 @@
...
+
+---
+apiVersion: openebs.io/v1alpha1
+kind: RunTask
+metadata:
+  name: upgrade-jiva-volume-0.8.2-0.9.0-get-sc-res-version
+  namespace: default
+spec:
+  meta: |
+    id: get
```
[the remark this answers is not in the document]
Comment (mittachaitu) [reply k8s/upgrades/0.8.2-0.9.0/jiva_upgrade_runtask.yaml]: Updated

**Pull request merged**